### PR TITLE
JavaScriptEngine common code rework

### PR DESCRIPTION
**Pull request by Tetraquark**

### CODE REVIEW

File: javascript/src/commonMain/kotlin/dev/icerock/moko/javascript/JavaScriptEvaluationException.kt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package dev.icerock.moko.javascript
```
Comment (Alex009) [inline javascript/src/commonMain/kotlin/dev/icerock/moko/javascript/JavaScriptEvaluationException.kt:5]: copyright please

File: javascript/src/commonMain/kotlin/dev/icerock/moko/javascript/JavaScriptEvaluationException.kt [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+
+class JavaScriptEvaluationException(
+    throwable: Throwable?,
```
Comment (Alex009) [inline javascript/src/commonMain/kotlin/dev/icerock/moko/javascript/JavaScriptEvaluationException.kt]: cause, as i think

**Review by Dorofeev — APPROVED** [review]

File: javascript/src/androidMain/kotlin/dev/icerock/moko/javascript/JavaScriptEngine.kt [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            }
+        }
+
+        return when (result) {
+            result == null -> JsType.Null
+            is Boolean -> JsType.Bool(result)
+            is Int -> JsType.IntNum(result)
+            is Double -> JsType.DoubleNum(result)
+            is Float -> JsType.DoubleNum(result.toDouble())
+            else -> throw IllegalStateException("Impossible JavaScriptEngine handler state")
```
Comment (Alex009) [inline javascript/src/androidMain/kotlin/dev/icerock/moko/javascript/JavaScriptEngine.kt]: need more information - write $result into exception message too

and maybe will use `JavaScriptEvaluationException` ?

Comment (Tetraquark) [reply javascript/src/androidMain/kotlin/dev/icerock/moko/javascript/JavaScriptEngine.kt]: Fixed

File: javascript/src/commonMain/kotlin/dev/icerock/moko/javascript/JavaScriptEvaluationException.kt (hunk 2 above)
Comment (Tetraquark) [reply javascript/src/commonMain/kotlin/dev/icerock/moko/javascript/JavaScriptEvaluationException.kt]: Fixed

File: javascript/src/androidAndroidTest/kotlin/dev/icerock/moko/javascript/JavaScriptEngineTest.kt [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        assertEquals(
+            expected = JsType.IntNum(19),
+            actual = javaScriptEngine.evaluate(context = context, script = "list[1]+number")
+        )
+
+        // FIXME: Falls with exception
+        // Script: var list = [5,15];var number = 4;var doubleString = " Hello ";doubleString+list[0]
+        // Exception:
+        // IllegalStateException: Reader has not consumed the whole input:
+        // JsonReader(source=' Hello 5', currentPosition=8, tokenClass=0, tokenPosition=7, offset=7)
```
Comment (Alex009) [inline javascript/src/androidAndroidTest/kotlin/dev/icerock/moko/javascript/JavaScriptEngineTest.kt]: maybe source of it in function `handleQuickJsResult` where `json.parseToJsonElement(result)` throw this exception?

File: javascript/src/androidMain/kotlin/dev/icerock/moko/javascript/JavaScriptEngine.kt [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        val scriptWithContext = convertContextMapToJsScript(context) + script
+        val result = quickJs.evaluate(scriptWithContext)
+        return handleQuickJsResult(result)
+    }
+
+    private fun convertContextMapToJsScript(context: Map<String, JsType>): String {
+        if (context.isEmpty()) return ""
+
+        return context.mapNotNull { pair ->
+            pair.value.value?.let { "var ${pair.key} = $it;" }
```
Comment (Alex009) [inline javascript/src/androidMain/kotlin/dev/icerock/moko/javascript/JavaScriptEngine.kt]: but how it set string? and how set json? for string i think we should wrapp into `"` and for json we should convert by `JSON.parse`

File: javascript/src/commonMain/kotlin/dev/icerock/moko/javascript/JsType.kt [hunk 6]
```diff
@@ -4,10 +4,36 @@
...
+    get() {
+        return when (this) {
+            is JsType.Bool -> value
+            is JsType.Str -> value
+            is JsType.IntNum -> value
+            is JsType.DoubleNum -> value
+            is JsType.Json -> value.toString()
+            is JsType.Null -> null
+        }
+    }
```
Comment (Alex009) [inline javascript/src/commonMain/kotlin/dev/icerock/moko/javascript/JsType.kt]: we can just move it into `JsType` class as base i think

File: javascript/src/iosMain/kotlin/dev/icerock/moko/javascript/JavaScriptEngine.kt [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+private fun JsType.getValue(): Any? {
+    return when (this) {
+        is JsType.Bool -> value
+        is JsType.Str -> value
+        is JsType.IntNum -> value
+        is JsType.DoubleNum -> value
+        is JsType.Json -> value.getValue()
+        is JsType.Null -> null
+    }
+}
```
Comment (Alex009) [inline javascript/src/iosMain/kotlin/dev/icerock/moko/javascript/JavaScriptEngine.kt]: duplicated - in common code also exist this

**Pull request merged**